teachertool/src/components/SplitPane.tsx

```diff
@@ -5,20 +5,115 @@ import { classList } from "react-common/components/util";
 interface IProps {
     className?: string;
     split: "horizontal" | "vertical";
-    defaultSize: number | string;
+    defaultSize: number | string; // The size to reset to when double clicking the splitter.
+    startingSize?: number | string; // The size to use initially when creating the splitter. Defaults to `defaultSize`.
     primary: "left" | "right";
     left: React.ReactNode;
     right: React.ReactNode;
+    leftMinSize?: number | string;
+    rightMinSize?: number | string;
+    onResizeEnd?: (size: number | string) => void;
 }
 
-export const SplitPane: React.FC<IProps> = ({ className, split, left, right }) => {
+export const SplitPane: React.FC<IProps> = ({
+    className,
+    split,
+    defaultSize,
+    startingSize,
+    left,
+    right,
+    leftMinSize,
+    rightMinSize,
+    onResizeEnd,
+}) => {
+    const [size, setSize] = React.useState(startingSize ?? defaultSize);
+    const [isResizing, setIsResizing] = React.useState(false);
+    const containerRef = React.useRef<HTMLDivElement>(null);
+
+    React.useEffect(() => {
+        if (!isResizing) {
+            onResizeEnd?.(size);
+        }
+    }, [isResizing]);
+
+    function handleResizeMouse(event: MouseEvent) {
+        handleResize(event.clientX, event.clientY);
+    }
+
+    function handleResizeTouch(event: TouchEvent) {
+        if (event.touches.length >= 1) {
+            handleResize(event.touches[0].clientX, event.touches[0].clientY);
+        }
+    }
+
+    function handleResize(clientX: number, clientY: number) {
+        const containerRect = containerRef.current?.getBoundingClientRect();
+        if (containerRect) {
+            const newSize =
+                split === "vertical"
+                    ? `${((clientX - containerRect.left) / containerRect.width) * 100}%`
+                    : `${((clientY - containerRect.top) / containerRect.height) * 100}%`;
+
+            setIsResizing(true); // Do this here rather than inside startResizing to prevent interference with double click detection.
```

Comment:
It probably doesn't make that much of a difference, but is there a reason you don't do this at the start of the `if` statement?

Reply:
I just thought it looked nice to keep the set-state calls together. Do you think we should move it up?

Reply:
My initial thought is yes, we should move it up because we want to make sure that `onResizeEnd` doesn't get called while the container is actively resizing, but I don't know if that case could happen or if it would be the worst thing if it did. It also might be the case that putting `setResizing` above the `newSize` definition would cause more problems with the `useEffect`. It's just one of those React things that I'm a bit fuzzy on.

Reply:
It's common to use ResizeObserver in these scenarios. Did you consider using ResizeObserver here before going with this solution?

Reply:
I moved it up just in case.

As for the ResizeObserver, I'm not sure that'd help much in this case. I think it's actually preferable to react to the mouse up/down/move events, which feel closer to user's intent than trying to deduce from resize observations.